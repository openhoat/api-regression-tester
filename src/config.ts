import { z } from 'zod'
import type { ConfigCustom } from './util/config-builder'

const schema = z.object({
  legacyBaseUrl: z.string().default('http://localhost:3000'),
  actualBaseUrl: z.string().default('http://localhost:3001'),
  logLevel: z
    .enum(['silent', 'fatal', 'error', 'warn', 'info', 'debug', 'trace'])
    .default('warn'),
  timeout: z.number().default(5000),
  modelProvider: z.enum(['openai', 'gemini', 'mistral', 'ollama']),
  model: z.string().optional(),
  googleApiKey: z.string().optional(),
  openaiApiKey: z.string().optional(),
  mistralApiKey: z.string().optional(),
  ollamaBaseUrl: z.string().optional(),
})

export type Config = z.infer<typeof schema>

const custom: ConfigCustom<Config> = {}

export const configSpec = { schema, custom }
