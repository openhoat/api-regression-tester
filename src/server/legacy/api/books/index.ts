import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify'
import { v4 as uuidv4 } from 'uuid'

export interface Book {
  id: string
  title: string
  author: string
  createdAt: number
}

const books: Book[] = [
  {
    id: uuidv4(),
    title: 'Notre-Dame de Paris',
    author: 'Victor Hugo',
    createdAt: 1550044806000,
  },
  {
    id: uuidv4(),
    title: 'Le Petit Prince',
    author: 'Antoine de Saint-ExupÃ©ry',
    createdAt: 1297584006000,
  },
]

export const booksRouter = async (fastify: FastifyInstance) => {
  fastify.get(
    '/',
    async (req: FastifyRequest<{ Querystring: { title?: string } }>) => {
      const { title } = req.query
      if (title) {
        return books.filter((book) => book.title === title)
      }
      return books
    },
  )
  fastify.get(
    '/:id',
    (
      req: FastifyRequest<{
        Params: { id: string }
      }>,
      reply: FastifyReply,
    ) => {
      const { id } = req.params
      const book = books.find((b) => b.id === id)
      if (!book) {
        reply.callNotFound()
        return
      }
      return book
    },
  )
}
