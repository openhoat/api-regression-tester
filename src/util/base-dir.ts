import { relative, resolve } from 'node:path'

const relativeBaseDir = '../..'
const thisScriptDir = __dirname
const currentDir = process.cwd()
const absoluteBaseDir = resolve(thisScriptDir, relativeBaseDir)

export const baseDir = relative(currentDir, absoluteBaseDir)
