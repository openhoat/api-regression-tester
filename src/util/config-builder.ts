import type { ZodType, z } from 'zod'
import { loadEnv } from './env.helper'
import { toPascalCase } from './string.helper'

export type ConfigCustomItem = {
  transform?: z.ZodType<unknown>
  envVarName?: string
}

export type ConfigCustom<T> = Partial<Record<keyof T, ConfigCustomItem>>

export type ConfigSpec<S extends Record<string, ZodType>, T> = {
  schema: z.ZodObject<S>
  custom?: ConfigCustom<T>
}

export type EnvProvider = Record<string, string | undefined>

export const buildConfig = <S extends Record<string, ZodType>, T>(
  { schema, custom }: ConfigSpec<S, T>,
  envProvider: EnvProvider = process.env,
) => {
  if (envProvider === process.env) {
    loadEnv(envProvider.ENVIRONMENT)
  }
  const { VERBOSE } = envProvider
  const verbose = VERBOSE === 'true'
  const configShemaEntries = Object.keys(schema.shape)
  const config = schema.parse(
    configShemaEntries.reduce(
      (acc, name: string) => {
        const spec = custom?.[name as keyof T]
        const envVarName = spec?.envVarName ?? toPascalCase(name)
        const envVarValue = envProvider[envVarName]
        if (envVarValue === undefined) {
          return acc
        }
        const transform = spec?.transform ?? schema.shape[name]
        acc[name] = transform.parse(envVarValue)
        return acc
      },
      {} as Record<string, unknown>,
    ),
  )
  if (verbose) {
    console.debug('Loaded config:', config)
  }
  return config
}
