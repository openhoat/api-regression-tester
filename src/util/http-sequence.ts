import { readFile } from 'node:fs/promises'
import { JSONPath } from 'jsonpath-plus'
import _ from 'lodash'
import { parse } from 'yaml'
import {
  type ApiRequest,
  type ApiResponse,
  type HttpMethod,
  sendRequest,
} from './http.helper'
import type { JsonRecord, JsonValue } from './types'

export type HttpSequenceStep = {
  name: string
  request: {
    method: HttpMethod
    endpoint: string
    basicAuth?: {
      username: string
      password: string
    }
    headers?: Record<string, string>
    queryParams?: Record<string, string>
    payload?: unknown
  }
  transformations?: {
    key: string
    type?: string
    value?: unknown
    args?: string[]
  }[]
  assertions?: { name: string; type: string; args?: string[] }[]
}

export type HttpSequence = {
  name: string
  baseUrl?: string
  steps: HttpSequenceStep[]
  responseMapping: Record<string, string>
}

export type ConversationStep = {
  request: ApiRequest
  response?: ApiResponse
}

export type Conversation = Record<string, ConversationStep>

export type HttpSequenceState<T extends JsonRecord> = {
  result?: T
} & JsonRecord

export type HttpSequenceResult<T extends JsonRecord> = {
  conversation: Conversation
  state: HttpSequenceState<T>
}

export type Asserter = (store: JsonRecord, ...args: unknown[]) => void

export type Asserters = Record<string, Asserter>

export type Transformer = (store: JsonRecord, ...args: unknown[]) => unknown

export type Transformers = Record<string, Transformer>

export type SequenceOptions<T extends JsonRecord> = Partial<{
  conversation: Conversation
  state: HttpSequenceState<T>
  asserters: Asserters
  transformers: Transformers
}>

export type SequenceStepOptions<T extends JsonRecord> = Partial<{
  baseUrl: string
}> &
  SequenceOptions<T>

export const evaluateValue = (
  value: string,
  store: JsonRecord,
  defaultValue: unknown = undefined,
) => {
  if (!value.startsWith('$')) {
    return defaultValue
  }
  const [computedValue] = JSONPath({ path: value, json: store })
  return computedValue
}

export const runSequenceStep = async <T extends JsonRecord>(
  step: HttpSequenceStep,
  options?: SequenceStepOptions<T>,
): Promise<HttpSequenceResult<T>> => {
  const { name, request, transformations, assertions } = step
  const {
    conversation = {} satisfies Conversation,
    state = {} as HttpSequenceState<T>,
    asserters,
    transformers,
  } = options ?? {}
  const store = { conversation, state } as unknown as JsonRecord
  const { endpoint, method, queryParams, headers, payload, basicAuth } = request
  if (headers) {
    for (const [key, value] of Object.entries(headers)) {
      headers[key] = `${evaluateValue(value, store, value)}`
    }
  }
  const apiRequest: ApiRequest = {
    method,
    path: endpoint,
    basicAuth,
    headers,
    params: queryParams,
    body: payload,
  }
  if (!conversation[name]) {
    conversation[name] = { request: apiRequest }
  }
  conversation[name].response = await sendRequest<JsonValue>(apiRequest, {
    baseUrl: options?.baseUrl,
  })
  if (transformations) {
    transformations.forEach(({ key, type, value, args = [] }) => {
      const transformer = type && transformers?.[type]
      if (type && !transformer) {
        throw new Error(`Missing transformer for type "${type}"`)
      }
      const finalValue = transformer
        ? transformer(store, ...args)
        : JSONPath({ path: `${value}`, json: store })
      _.set(state, `result.${key}`, finalValue)
    })
  }
  if (assertions) {
    assertions.forEach(({ type, args = [] }) => {
      const asserter = asserters?.[type]
      if (!asserter) {
        throw new Error(`Missing asserter for type "${type}"`)
      }
      asserter(store, ...args)
    })
  }
  return store as unknown as { conversation: Conversation; state: T }
}

export const runSequence = async <T extends JsonRecord>(
  sequence: HttpSequence,
  options?: SequenceOptions<T>,
): Promise<HttpSequenceResult<T>> => {
  const { baseUrl, steps } = sequence
  const {
    conversation = {} satisfies Conversation,
    state = {} as HttpSequenceState<T>,
    asserters,
    transformers,
  } = options ?? {}
  const store = { conversation, state }
  const promise = steps.reduce(
    async (acc: Promise<HttpSequenceResult<T>>, step) => {
      const { conversation, state } = await acc
      return runSequenceStep(step, {
        baseUrl,
        conversation,
        state,
        asserters,
        transformers,
      })
    },
    Promise.resolve(store),
  )
  await promise
  return store
}

export const loadYamlFile = async <T>(file: string): Promise<T> => {
  const content = await readFile(file, 'utf8')
  return parse(content) satisfies T
}
