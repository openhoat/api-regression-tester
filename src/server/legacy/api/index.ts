import type { FastifyInstance } from 'fastify'
import { bookingRouter } from './booking'
import { booksRouter } from './books'

export const apiRouter = async (fastify: FastifyInstance) => {
  fastify.register(bookingRouter, { prefix: '/booking' })
  fastify.register(booksRouter, { prefix: '/books' })
}
