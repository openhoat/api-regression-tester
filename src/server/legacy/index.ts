import Fastify from 'fastify'
import type { LevelWithSilentOrString } from 'pino'
import type { Service } from '../../util/service'
import { apiRouter } from './api'

export const createServer = (options?: {
  port?: number
  logLevel?: LevelWithSilentOrString
}): Service => {
  const { port, logLevel } = options ?? {}
  const app = Fastify({
    logger: {
      transport: {
        target: 'pino-pretty',
        options: {
          translateTime: 'HH:MM:ss Z',
          ignore: 'pid,hostname',
        },
      },
      level: logLevel,
    },
  })
  app.register(apiRouter, { prefix: '/api' })
  return {
    start: async () => {
      await app.listen({ port: port ?? 3000, host: '0.0.0.0' })
    },
    stop: async () => {
      await app.close()
    },
  }
}
