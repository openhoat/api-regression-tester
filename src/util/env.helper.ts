import { join } from 'node:path'
import { config as dotenvConfig } from 'dotenv'
import { baseDir } from './base-dir'

export const loadDotenvFile = (envName?: string) => {
  const envFilename = envName ? `.env.${envName}` : '.env'
  dotenvConfig({ path: join(baseDir, envFilename), quiet: true })
  const { VERBOSE } = process.env
  const verbose = VERBOSE === 'true'
  if (verbose) {
    console.debug(`Environment loaded from ${envFilename}`)
  }
}

export const loadEnv = (envName?: string) => {
  loadDotenvFile('local')
  if (envName && envName !== 'local') {
    loadDotenvFile(envName)
  }
  loadDotenvFile()
}
