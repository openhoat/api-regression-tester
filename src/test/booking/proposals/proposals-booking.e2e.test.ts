import { join } from 'node:path'
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals'
import { configSpec } from '../../../config'
import { createServer as createActualServer } from '../../../server/actual'
import { createServer as createLegacyServer } from '../../../server/legacy'
import { buildConfig } from '../../../util/config-builder'
import {
  type HttpSequence,
  loadYamlFile,
  runSequence,
} from '../../../util/http-sequence'
import type { Service } from '../../../util/service'
import { mockApis } from '../../books/api-mocks'
import { anyStringArray } from '../../util/test.helper'
import { asserters } from '../util/asserters'
import { transformers } from '../util/transformers'

const { legacyBaseUrl, actualBaseUrl, logLevel } = buildConfig(configSpec)

const { MOCK, LOCAL_API, VERBOSE } = process.env
const verbose = VERBOSE === 'true'
const mock = MOCK === 'true'
const localApi = LOCAL_API === 'true'

describe('Booking API e2e test', () => {
  const servers: {
    legacy: Service | null
    actual: Service | null
  } = { legacy: null, actual: null }
  beforeAll(async () => {
    if (localApi) {
      const legacyServer = createLegacyServer({ logLevel })
      const actualServer = createActualServer({ logLevel })
      await Promise.all([legacyServer.start(), actualServer.start()])
      servers.legacy = legacyServer
      servers.actual = actualServer
    }
    if (mock) {
      mockApis(legacyBaseUrl, actualBaseUrl)
    }
  })
  afterAll(async () => {
    const { legacy: legacyServer, actual: actualServer } = servers
    if (legacyServer) {
      await legacyServer.stop()
      servers.legacy = null
    }
    if (actualServer) {
      await actualServer.stop()
      servers.actual = null
    }
  })
  describe('proposals sequence', () => {
    let sequence: HttpSequence
    beforeAll(async () => {
      sequence = await loadYamlFile(
        join(__dirname, 'proposals-booking.sequence.yml'),
      )
    })
    test('should execute proposals sequence', async () => {
      // When
      type ApiResult = {
        legacyOriginLocalityIds: string[]
        legacyDestinationLocalityIds: string[]
        legacySegmentOriginLocalityIds: string[]
        legacySegmentDestinationLocalityIds: string[]
        legacyOriginLocalityCodes: string[]
        legacyDestinationLocalityCodes: string[]
        legacyPrices: number[]
      }
      const store = await runSequence<ApiResult>(sequence, {
        transformers,
        asserters,
      })
      // Then
      if (verbose) {
        console.log('store state:', JSON.stringify(store.state, null, 2))
      }
      const { result } = store.state
      expect(result).toBeTruthy()
      expect(result).toMatchSnapshot({
        legacyOriginLocalityIds: anyStringArray(
          result?.legacyOriginLocalityIds.length,
        ),
        legacyDestinationLocalityIds: anyStringArray(
          result?.legacyDestinationLocalityIds.length,
        ),
        legacySegmentOriginLocalityIds: anyStringArray(
          result?.legacySegmentOriginLocalityIds.length,
        ),
        legacySegmentDestinationLocalityIds: anyStringArray(
          result?.legacySegmentDestinationLocalityIds.length,
        ),
      })
    })
  })
})
