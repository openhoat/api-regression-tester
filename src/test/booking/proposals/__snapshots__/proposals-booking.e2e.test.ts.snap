// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Booking API e2e test proposals sequence should execute proposals sequence 1`] = `
{
  "actualDestinationLocalityCodes": [
    "FRUIP",
    "FRUIP",
    "FRUIP",
    "FRUIP",
    "FRUIP",
    "FRUIP",
  ],
  "actualOriginLocalityCodes": [
    "FRPMO",
    "FRPMO",
    "FRPMO",
    "FRPMO",
    "FRPMO",
    "FRPMO",
  ],
  "actualPrices": [
    47,
    29,
    76,
    76,
    76,
    76,
  ],
  "legacyDestinationLocalityCodes": [
    "FRUIP",
    "FRUIP",
    "FRUIP",
    "FRUIP",
    "FRUIP",
    "FRUIP",
  ],
  "legacyDestinationLocalityIds": [
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
  ],
  "legacyOriginLocalityCodes": [
    "FRPMO",
    "FRPMO",
    "FRPMO",
    "FRPMO",
    "FRPMO",
    "FRPMO",
  ],
  "legacyOriginLocalityIds": [
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
  ],
  "legacyPrices": [
    47,
    29,
    76,
    76,
    76,
    76,
  ],
  "legacySegmentDestinationLocalityIds": [
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
  ],
  "legacySegmentOriginLocalityIds": [
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
    Any<String>,
  ],
}
`;
